<p style='text-align:center; font-size: 36px; color:#0B0B23; font-weight: bold;'>Mean-Variance-Stardard Deviation Calculator</p>

This notebook/project creates a python function to show the mean, variance, standard deviation, max, min and sum of the rows, columns and elements in a 3x3 matrix. The function recieves a list as parameter that contains 9 digits and creates the matrix itself.

### Let's start

First off, let's import numpy

In [1]:
import numpy as np

Now let's write the function

In [2]:
def calculate(list):

    # Check if the parameters contain nine numbers
    if len(list) != 9:
        raise ValueError('List must contain nine numbers.')

    list = np.array(list)
    list = list.reshape((3,3))
    calculations = {
        'mean': [
            list.mean(axis=0).tolist(),
            list.mean(axis=1).tolist(),
            list.mean().tolist()
        ],
        'variance': [
            list.var(axis=0).tolist(),
            list.var(axis=1).tolist(),
            list.var().tolist()
        ],
        'standard deviation': [
            list.std(axis=0).tolist(),
            list.std(axis=1).tolist(),
            list.std().tolist()
        ],
        'max': [
            list.max(axis=0).tolist(),
            list.max(axis=1).tolist(),
            list.max().tolist()
        ],
        'min': [
            list.min(axis=0).tolist(),
            list.min(axis=1).tolist(),
            list.min().tolist()
        ],
        'sum': [
            list.sum(axis=0).tolist(),
            list.sum(axis=1).tolist(),
            list.sum().tolist()
        ]
    }

    return calculations

### Let's test it

Trying to pass more or less numbers

In [3]:
wrong_list = [0,1,2,3,4,5,6,7,8,9]
calculate(wrong_list)

ValueError: List must contain nine numbers.

In [4]:
wrong_list.clear()
calculate(wrong_list)

ValueError: List must contain nine numbers.

Now test if actually works

In [5]:
right_list = [0,1,2,3,4,5,6,7,8]
calculate(right_list)

{'mean': [[3.0, 4.0, 5.0], [1.0, 4.0, 7.0], 4.0],
 'variance': [[6.0, 6.0, 6.0],
  [0.6666666666666666, 0.6666666666666666, 0.6666666666666666],
  6.666666666666667],
 'standard deviation': [[2.449489742783178,
   2.449489742783178,
   2.449489742783178],
  [0.816496580927726, 0.816496580927726, 0.816496580927726],
  2.581988897471611],
 'max': [[6, 7, 8], [2, 5, 8], 8],
 'min': [[0, 1, 2], [0, 3, 6], 0],
 'sum': [[9, 12, 15], [3, 12, 21], 36]}

### Let's analyze if it's right

<hr>

It turned our 2d array into a 3 dimension one:

[0, 1, 2]
[3, 4, 5]
[6, 7, 8]

<hr>

Mean in the axis 0 (columns):

col 1 = [0, 3, 6]; mean = 3
col 2 = [1, 4, 7]; mean = 4
col 3 = [2, 5, 8]; mean = 5

Mean in the axis 1 (rows):

row 1 = [0, 1, 2]; mean = 1
row 2 = [3, 4, 5]; mean = 4
row 3 = [6, 7, 8]; mean = 7

Mean of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; mean = 4

<hr>

Variance in the axis 0 (columns):

col 1 = [0, 3, 6]; variance = 6
col 2 = [1, 4, 7]; variance = 6
col 3 = [2, 5, 8]; variance = 6

Variance in the axis 1 (rows):

row 1 = [0, 1, 2]; variance = 0.66
row 2 = [3, 4, 5]; variance = 0.66
row 3 = [6, 7, 8]; variance = 0.66

Variance of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; variance = 6.66

<hr>

Standard Deviation in the axis 0 (columns):

col 1 = [0, 3, 6]; std = 2.45
col 2 = [1, 4, 7]; std = 2.45
col 3 = [2, 5, 8]; std = 2.45

Standard Deviation in the axis 1 (rows):

row 1 = [0, 1, 2]; std = 0.82
row 2 = [3, 4, 5]; std = 0.82
row 3 = [6, 7, 8]; std = 0.82

Standard Deviation of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; std = 2.60

<hr>

Highest number in the axis 0 (columns):

col 1 = [0, 3, 6]; max = 6
col 2 = [1, 4, 7]; max = 7
col 3 = [2, 5, 8]; max = 8

Highest number in the axis 1 (rows):

row 1 = [0, 1, 2]; max = 2
row 2 = [3, 4, 5]; max = 5
row 3 = [6, 7, 8]; max = 8

Highest of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; max = 8

<hr>

Lowest number in the axis 0 (columns):

col 1 = [0, 3, 6]; min = 0
col 2 = [1, 4, 7]; min = 1
col 3 = [2, 5, 8]; min = 2

Lowest number in the axis 1 (rows):

row 1 = [0, 1, 2]; min = 0
row 2 = [3, 4, 5]; min = 3
row 3 = [6, 7, 8]; min = 6

Lowest of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; min = 0

<hr>

Sum of all numbers in the axis 0 (columns):

col 1 = [0, 3, 6]; sum = 9
col 2 = [1, 4, 7]; sum = 12
col 3 = [2, 5, 8]; sum = 15

Sum of all numbers in the axis 1 (rows):

row 1 = [0, 1, 2]; sum = 3
row 2 = [3, 4, 5]; sum = 12
row 3 = [6, 7, 8]; sum = 21

Sum of all numbers in the matrix:

list = [0, 1, 2, 3, 4, 5, 6, 7, 8]; sum = 36

<hr>

### It's Right!